File: shpam.py
from subprocess import check_output
from hashlib import sha1

LOC = '.'
CONFORMITY = 6

info = check_output(['git', '--no-replace-objects', 'cat-file', 'commit', 'HEAD'], cwd=LOC)

validChars = ''.join(set('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()[]{}<>$@#"\'\\+-*/~&!:;?%'))
mx = len(validChars)-1

c = [0]

def s():
    return ''.join(map(lambda i:validChars[i],c))

def ind(i):
    global c
    if c[i] == mx:
        c[i] = 0
        if i == len(c)-1:
            c.append(0)
        else:
            ind(i+1)
    else:
        c[i] += 1

def test():
    v = info[:-1] + bytes(s(),'utf-8') + b'\n'
    sh = sha1()
    sh.update(bytes(f'commit {len(v)}\0', 'ascii')+v)
    return ''.join(map(lambda c:hex(c)[2:].ljust(2,'0'),sh.digest()))

while 1:
    t = test()
    nz = len(t)-len(t.lstrip('0'))
    if nz >= CONFORMITY:
        t = 'CONFORMS'
        if nz > CONFORMITY:
            t += ' +'+str(nz-CONFORMITY)
        print(t)
        print('\x1b[92;1m'+(b'\n'.join(info.splitlines()[4:])).decode('utf-8')+s()+'\x1b[39;22m')
        print('END\n')
    ind(0)

File: shpam.js
const cp = require('node:child_process');
const { createHash } = require('node:crypto');

const LOC = '.';
const CONFORMITY = 6;

const info = cp.execFileSync('git', ['--no-replace-objects', 'cat-file', 'commit', 'HEAD'], {cwd:LOC}).toString('utf-8');

const validChars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()[]{}<>$@#"\'\\+-*/~&!:;?%';
const mx = validChars.length-1;

const c = [0];

const s = () => c.map(i=>validChars[i]).join('');

function ind(i) {
    if (c[i] == mx) {
        c[i] = 0;
        if (i == c.length-1)
            c.push(0);
        else
            ind(i+1);
    }
    else
        c[i]++;
}

function test() {
    const v = info.slice(0,-1) + s() + '\n';
    return createHash('sha1').update(`commit ${v.length}\0${v}`).digest('hex');
}

for (;;) {
    const t = test();
    nz = /^0*/.exec(t)[0].length;
    if (nz >= CONFORMITY) {
        console.log(`CONFORMS${nz>CONFORMITY?' +'+(nz-CONFORMITY):'\n\x1b[92;1m'}${info.split(/\r?\n/g).slice(5,-1).join('\n')+s()}\x1b[39;22m\nEND\n`);
    }
    ind(0);
}
